ipam/allocator.go

```diff
@@ -159,97 +187,293 @@ const (
+		if err != nil {
+			return "", nil, nil, err
+		}
+		if pw, err = adjustAndCheckSubnetSize(nw); err != nil {
+			return "", nil, nil, err
+		}
 	}
+
+	k := subnetKey{addressSpace: addressSpace, subnet: pool, childSubnet: subPool}
+
 retry:
-	if a.contains(addrSpace, subnetInfo) {
-		return ErrOverlapSubnet
+	if ipr == nil {
+		a.Lock()
+		if p, ok = a.subnets[k]; ok {
+			p.RefCount++
+			goto finalize
+		}
+		a.Unlock()
+		if a.contains(addressSpace, nw) {
+			return "", nil, nil, ipamapi.ErrPoolOverlap
+		}
 	}
 
-	// Store the configured subnet and sync to datatstore
-	key := subnetKey{addrSpace, subnetInfo.Subnet.String(), ""}
 	a.Lock()
-	a.subnets[key] = subnetInfo
+	if p, ok = a.subnets[k]; ok {
+		c = p
+		for ok {
+			c.RefCount++
+			c, ok = a.subnets[c.ParentKey]
+		}
+		goto finalize
+	}
+
+	p = &PoolData{
+		Pool:     pw,
+		Range:    ipr,
+		RefCount: 1,
+	}
+
+	if p.Range == nil {
+		a.Unlock()
+		if err := a.insertBitMask(k, p.Pool); err != nil {
+			return "", nil, nil, err
+		}
+		a.Lock()
+		a.subnets[k] = p
+		goto finalize
+	}
+
+	p.ParentKey = subnetKey{addressSpace: addressSpace, subnet: pool}
+	a.subnets[k] = p
+	if pp, ok = a.subnets[p.ParentKey]; ok {
+		c = pp
+		for ok {
+			c.RefCount++
+			c, ok = a.subnets[c.ParentKey]
+		}
+		goto finalize
+	}
+
+	a.Unlock()
+
+	// parent does not exist
+	pp = &PoolData{
+		Pool:     p.Pool,
+		RefCount: 1,
+	}
+	if err := a.insertBitMask(p.ParentKey, pp.Pool); err != nil {
+		return "", nil, nil, err
+	}
+
+	a.Lock()
+	a.subnets[p.ParentKey] = pp
+finalize:
 	a.Unlock()
 	err = a.writeToStore()
 	if err != nil {
 		if _, ok := err.(types.RetryError); !ok {
-			return types.InternalErrorf("subnet configuration failed because of %s", err.Error())
+			return "", nil, nil, types.InternalErrorf("pool configuration failed because of %s", err.Error())
 		}
 		// Update to latest
 		if erru := a.readFromStore(); erru != nil {
-			// Restore and bail out
-			a.Lock()
-			delete(a.addresses, key)
-			a.Unlock()
-			return fmt.Errorf("failed to get updated subnets config from datastore (%v) after (%v)", erru, err)
+			return "", nil, nil, fmt.Errorf("failed to get updated pool config from datastore (%v) after (%v)", erru, err)
 		}
 		goto retry
 	}
 
-	// Insert respective bitmasks for this subnet
-	a.insertAddressMasks(key, subnetList)
+	return k.String(), p.Pool, nil, nil
+}
+
+func (a *Allocator) insertBitMask(key subnetKey, pool *net.IPNet) error {
+	ipVer := getAddressVersion(pool.IP)
+	ones, bits := pool.Mask.Size()
+	numAddresses := uint32(1 << uint(bits-ones))
+
+	if ipVer == v4 {
+		// Do not let broadcast address be reserved
+		numAddresses--
+	}
+
+	// Generate the new address masks. AddressMask content may come from datastore
+	h, err := bitseq.NewHandle(dsDataKey, a.store, key.String(), numAddresses)
+	if err != nil {
+		return err
+	}
+
+	if ipVer == v4 {
+		// Do not let network identifier address be reserved
+		h.Set(0)
```

Comment:
I am not sure I understand your comment.
`insertBitMask` is done only for the parent pool.